docs/rfc/007-endpoint-errors.md

```diff
@@ -0,0 +1,229 @@
+    default-auth: header
+    endpoints:
+      getSomeResult:
+        http: GET /
+        returns: list<rid>
+        errors:
+          - BlueMoon
+        docs: Returns some result, but once in a blue moon throws an error
+```
+
+
+**Undertow Service**
+
+```
+@Generated("com.palantir.conjure.java.services.UndertowServiceInterfaceGenerator")
+public interface MyConjureService {
+    /**
+     * Returns some result, but once in a blue moon throws an error
+     * @apiNote {@code GET /}
+     */
+    List<ResourceIdentifier> getSomeResult(AuthHeader authHeader)
+            throws MyConjureServiceErrors.BlueMoonServiceException;
+```
+
+**Undertow Resource**
+
+```
+public final class MyConjureResource implements MyConjureService {
+
+   ...
+
+   @Override
+   public List<ResourceIdentifier> getSomeResult(AuthHeader authHeader)
+          throws BlueMoonServiceException {
+      throw MyConjureServiceErrors.throwIfBlueMoon(isBlueMoon);
+   }
+```
+
+### Example Dialogue Client Implementation via Exceptions
+
+
+```
+MyConjureServiceBlocking myConjureService = 
+    witchcraft.conjureClients()
+        .client(MyConjureServiceBlocking.class, "my-conjure-service").get();
+
+try {
+    myConjureService.getSomeResult(AuthHeader.valueOf("authHeader"));
+} catch (BlueMoonRemoteException e) { // <- checked exception so has to be caught!
+    handleBlueMoonEvent();
+}
+```
+
+### Example Dialogue Client Implementation via Union Types
+
+```
+MyConjureServiceBlocking myConjureService = 
+    witchcraft.conjureClients()
+        .client(MyConjureServiceBlocking.class, "my-conjure-service").get();
+
+List<ResourceIdentifier> myUnsafeResult = myConjureService.getSomeResult(AuthHeader.valueOf("authHeader"));
+
+// v- is this null in the failure case?
+List<ResourceIdentifier> mySafeResult1 = myConjureService.getSomeResult(AuthHeader.valueOf("authHeader"), MyConjureServiceErrors.getSomeResultBuilder()
+    .blueMoon(e -> handleBlueMoonEvent())
+    .build());
+    
+// v- could wrap the response into { ok: boolean, result: List<ResourceIdentifier> }, which throws a runtime exception
+//    if you try to access `result` but `ok == false`.
+SomeResultResponse mySafeResult2 = myConjureService.getSomeResult(AuthHeader.valueOf("authHeader"), MyConjureServiceErrors.getSomeResultBuilder()
+    .blueMoon(e -> handleBlueMoonEvent())
+    .build());
+    
+// v- alternative implementation where you force logic splitting
+myConjureService.getSomeResult(AuthHeader.valueOf("authHeader"), MyConjureServiceErrors.getSomeResultBuilder()
+    .blueMoon(e -> handleBlueMoonEvent())
+    .ok(mySafeResult3 -> handleOk(mySafeResult3))
+    .build());
+    
+// v- or via `Visitor<T>` and return `T` in all branches
+SomeOtherResult otherResult = myConjureService.getSomeResult(AuthHeader.valueOf("authHeader"), MyConjureServiceErrors.getSomeResultBuilder()
+    .blueMoon(e -> handleBlueMoonEvent())
+    .ok(mySafeResult3 -> handleOk(mySafeResult3))
+    .build());
```

Comment:
This style visitor was used as an example in the original discussion for why folks did not want to simply return a union from their endpoint.

Reply:
My intention here was purely as the 'shell' around the existing error mechanism - so we still ship the same errors across the wire (so the JSON is identical to both existing and the exceptions-based mechanism) but we just shepherd the result into a union rather than throwing it as an exception.

Reply:
I'd prefer returning a result type here, to allow for scoped handling:

```
// Lazy
myConjureService.getSomeResult(AuthHeader.valueOf("authHeader")).orThrow();

// Cleaner
myConjureService.getSomeResult(AuthHeader.valueOf("authHeader")).handle(success => ..., failure => ...);

// Stream which drops failures
myList.stream()
    .map(entry => myConjureService.getSomeResult(AuthHeader.valueOf("authHeader"), entry))
    .flatMap(Result::successful)
    .collect(Collectors.toList())

// Stream which handles failures
myList.stream()
    .map(entry => myConjureService.getSomeResult(AuthHeader.valueOf("authHeader"), entry))
    .map(result => result.or(MyConjureServiceErrors.getSomeResultBuilder()
        .blueMoon(e -> deriveBackupValue(e))
        .build()))
    .collect(Collectors.toList())

```